### ISSUE TITLE
Consider supporting corepack and making pnpm the only package manager

### ISSUE THREAD

**Issue by markcowl** [microsoft/typespec#5196]
Following up from this contributed PR:https://github.com/microsoft/typespec/pull/5194

Right now, corepack is considered experimental, so cannot immediately take a dependency without some research.  Need to determine any implications of using pnpm everywhere, engineering impacts of switching to corepack, like how we would manage corepack security updates, and if we should take a dependency on an experimental feature of node.

**Comment by ncaq** [comment]
Since the project already depends on corepack, I did not think that would be a problem.

https://github.com/microsoft/typespec/blob/98b8ba4a5cc95b79c1487254dd0a9a44eccf6763/package.json#L5

https://github.com/microsoft/typespec/blob/98b8ba4a5cc95b79c1487254dd0a9a44eccf6763/eng/common/pipelines/ci.yml#L28

https://github.com/microsoft/typespec/blob/98b8ba4a5cc95b79c1487254dd0a9a44eccf6763/eng/tsp-core/pipelines/templates/install.yml#L30

https://github.com/microsoft/typespec/blob/98b8ba4a5cc95b79c1487254dd0a9a44eccf6763/docker/Dockerfile#L11

https://github.com/microsoft/typespec/blob/98b8ba4a5cc95b79c1487254dd0a9a44eccf6763/.devcontainer/Dockerfile#L3

**Comment by MyNameIsNeXTSTEP** [comment]
I understand the _"Consider supporting corepack"_ part, but let me ask you why the second part of this issue suppose to _"making pnpm the only package manager"_ ?

What's wrong with yarn, for example, why not let end users decide and configure which package manager to use, if considering supporting the corepack at the first place ?

**Comment by ncaq** [comment]
The end-user's story of npm install from the outside, for example, is not relevant here.
End users are,
`npm install @typespec/compiler`, `yarn add @typespec/compiler`, or `pnpm add @typespec/compiler`, whatever you like.
In fact, I personally prefer yarn berry over pnpm.

The problem in this case is that I have the project set up to use pnpm for internal dependency building, builds and tests, but npm is also called in the mix, which is prone to errors and confusion in the corepack environment.
This choice of package manager is an internal developer issue.
Node.js package managers are switchable to some extent, and corepack can manage which package manager to use for each project, so I prefer to use the package manager configured in the typespec project, independent of the other projects.
That's the ideology of corepack, so I'm assuming that those who have corepack installed will agree to swap it out on a per-project basis.
Whether that is npm, yarn, or pnpm is not a big issue.
The problem is that they are mixed.
Mixed is worse, so if you can get rid of the mix, I'd be fine with just npm.
I think pnpm is excellent software, but I would rather have a unified “bad” package manager than a mixed worst case scenario.
And npm is not that “bad” recently.

As I wrote in the PR commit text, this is not directly related to corepack.
The use of mixed package managers is confusing and should be unified as much as possible.

**Comment by timotheeguerin** [comment]
Yeah this is just for our internal build system. Any package manager that install in `node_modules` should be supported.

To resolve this we just need to replace all `npm run` with `pnpm run`
